### PR TITLE
feat: draft pr for execute root fragment at FE instead of CN

### PR DESCRIPTION
**Pull request by BowenXiao1999**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

Previously, the root fragment task is executed in CN instead of FE. This PR changes that by always yield the root fragment executor stream in FE. So we can reduce some network cost.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

## Refer to a related PR or issue link (optional)
solve #3269

### CODE REVIEW

**Comment by BowenXiao1999** [comment]
While debugging this pr, I found madsim test with more than 100 times still succeed, while the normal e2e test can reproduce this in 50% chance.

madsim: 
```
MADSIM_TEST_NUM=100 ./risedev sslt -- "e2e_test/batch/basic/func.slt.part"
```

normal e2e
```
./risedev p

# Try twice or more and will fail in my PC
./risedev slt -p 4566 -d dev './e2e_test/batch/basic/func.slt.part'
```

Published this out for possibilities for investigating potential bug in madsim, cc @wangrunji0408 . Feel free to try this out and hope this help. 

I will continue to debug this PR until the problems in normal e2e is solved, and will not amend the reproduced commit hash.

File: src/frontend/src/scheduler/mod.rs [hunk 1]
```diff
@@ -54,4 +57,8 @@ impl ExecutionContext {
     pub fn session(&self) -> &SessionImpl {
         &self.session
     }
+
+    pub fn to_batch_task(&self) -> FrontendBatchTaskContext {
```
Comment (liurenjie1024) [inline src/frontend/src/scheduler/mod.rs:61]:
```suggestion
    pub fn to_batch_task_context(&self) -> FrontendBatchTaskContext {
```

File: src/frontend/src/scheduler/distributed/query_manager.rs [hunk 2]
```diff
@@ -121,13 +126,17 @@ impl QueryResultFetcher {
         task_output_id: TaskOutputId,
         task_host: HostAddress,
         compute_client_pool: ComputeClientPoolRef,
+        root_fragment: Option<PlanFragment>,
```
Comment (liurenjie1024) [inline src/frontend/src/scheduler/distributed/query_manager.rs:129]: Why this is `Option`?

File: src/frontend/src/scheduler/distributed/query.rs [hunk 3]
```diff
@@ -374,6 +384,85 @@ impl QueryRunner {
...
+        PlanFragment {
+            root: Some(plan_node_prost),
+            exchange_info: Some(exchange_info),
+        }
+    }
+
+    fn convert_plan_node(
+        stage_execution: &StageExecution,
+        execution_plan_node: &ExecutionPlanNode,
+        task_id: TaskId,
```
Comment (liurenjie1024) [inline src/frontend/src/scheduler/distributed/query.rs:403]: These codes are copied from `StageExecution`?

**Review by liurenjie1024 — COMMENTED** [review]
Generally LGTM.

**Comment by liurenjie1024** [comment]
Here is another potential design:
1. Add some logic in `StageExecution`, e.g. if it's root stage, we create an async to run its plan fragment
2. Modify `QueryResultFetcher` to accept a channel rahter rpc 

This way we don't need to do much modify scheduling logic and avoid code repeating.

File: src/frontend/src/scheduler/distributed/query.rs (hunk 3 above)
Comment (BowenXiao1999) [reply src/frontend/src/scheduler/distributed/query.rs:403]: I think you mean LocalQueryExecution? 
Yes the code are copied, basically extract the root fragment from plan. i remembered that there are some logic different so no-reuse

File: src/frontend/src/scheduler/distributed/query_manager.rs (hunk 2 above)
Comment (BowenXiao1999) [reply src/frontend/src/scheduler/distributed/query_manager.rs:129]: Fixed

**Comment by BowenXiao1999** [comment]
> Here is another potential design:
> 
> 1. Add some logic in `StageExecution`, e.g. if it's root stage, we create an async to run its plan fragment
> 2. Modify `QueryResultFetcher` to accept a channel rahter rpc
> 
> This way we don't need to do much modify scheduling logic and avoid code repeating.

My understanding:
1. in `StageRunner::run`, if we detect root fragment, then we should treat it differently with other fragments: execute locally and write results into a channel.
2. QueryResultFetcher get the channel receiver, poll results one by one.

Do you mean that? I think LGTM

**Comment by liurenjie1024** [comment]
> > Here is another potential design:
> > 
> > 1. Add some logic in `StageExecution`, e.g. if it's root stage, we create an async to run its plan fragment
> > 2. Modify `QueryResultFetcher` to accept a channel rahter rpc
> > 
> > This way we don't need to do much modify scheduling logic and avoid code repeating.
> 
> My understanding:
> 
> 1. in `StageRunner::run`, if we detect root fragment, then we should treat it differently with other fragments: execute locally and write results into a channel.
> 2. QueryResultFetcher get the channel receiver, poll results one by one.
> 
> Do you mean that? I think LGTM

Yes

**Comment by BowenXiao1999** [comment]
> Here is another potential design:
Add some logic in StageExecution, e.g. if it's root stage, we create an async to run its plan fragment
Modify QueryResultFetcher to accept a channel rahter rpc
This way we don't need to do much modify scheduling logic and avoid code repeating.

Implemented

**Review by liurenjie1024 — APPROVED** [review]
LGTM

**Pull request merged**